File: ts/src/bootstrap/renderer.ts
/** Environment snapshot prompt rendering (AC-503). */

import type { EnvironmentSnapshot } from "./snapshot.js";

export function renderPromptSection(snapshot: EnvironmentSnapshot): string {
  const lines: string[] = ["## Environment"];

  const coreParts = [
    `Python ${snapshot.pythonVersion}`,
    `${snapshot.osName} ${snapshot.osVersion}`,
    snapshot.shell?.split("/").pop() ?? "",
    snapshot.cpuCount ? `${snapshot.cpuCount} CPU` : "",
    snapshot.memoryTotalMb ? `${snapshot.memoryTotalMb}MB RAM` : "",
    snapshot.diskFreeGb ? `${snapshot.diskFreeGb}GB free` : "",
  ].filter(Boolean);
  lines.push(coreParts.join(" | "));

  if (snapshot.gitBranch) {
    const dirty = snapshot.gitDirty ? ", dirty" : ", clean";
    const wt = snapshot.gitWorktree ? ", worktree" : "";
    const commit = snapshot.gitCommit
      ? ` (${snapshot.gitCommit}${dirty}${wt})`
      : "";
    lines.push(`Git: ${snapshot.gitBranch}${commit}`);
  }

  if (Object.keys(snapshot.availableRuntimes).length > 0) {
    const parts = Object.entries(snapshot.availableRuntimes)
      .sort(([a], [b]) => a.localeCompare(b))
      .map(([name, ver]) => `${name} ${ver}`);
    lines.push(`Runtimes: ${parts.join(", ")}`);
  }

  if (snapshot.notableFiles.length > 0) {
    const top = snapshot.notableFiles.slice(0, 8);
    const extras =
      snapshot.fileCount || snapshot.directoryCount
        ? ` (${snapshot.fileCount} files, ${snapshot.directoryCount} dirs)`
        : "";
    lines.push(`Notable: ${top.join(", ")}${extras}`);
  }

  if (snapshot.installedPackages.length > 0) {
    const n = snapshot.installedPackages.length;
    const lockNote =
      snapshot.lockfilesFound.length > 0
        ? ` (${snapshot.lockfilesFound.join(", ")})`
        : "";
    lines.push(`Packages: ${n} top-level${lockNote}`);
  }

  return lines.join("\n");
}

export function renderFullJson(snapshot: EnvironmentSnapshot): string {
  return JSON.stringify(snapshot, null, 2);
}

File: autocontext/src/autocontext/bootstrap/renderer.py
"""Environment snapshot prompt rendering (AC-503)."""

from __future__ import annotations

import json

from autocontext.bootstrap.snapshot import EnvironmentSnapshot


def render_prompt_section(snapshot: EnvironmentSnapshot) -> str:
    """Render a compact markdown section for prompt injection (~300-500 chars)."""
    lines: list[str] = ["## Environment"]

    # Core line: Python version | OS | shell | CPU | RAM | disk
    core_parts = [
        f"Python {snapshot.python_version}",
        f"{snapshot.os_name} {snapshot.os_version}",
        snapshot.shell.rsplit("/", 1)[-1] if snapshot.shell else "",
        f"{snapshot.cpu_count} CPU" if snapshot.cpu_count else "",
        f"{snapshot.memory_total_mb}MB RAM" if snapshot.memory_total_mb else "",
        f"{snapshot.disk_free_gb}GB free" if snapshot.disk_free_gb else "",
    ]
    lines.append(" | ".join(p for p in core_parts if p))

    # Git
    if snapshot.git_branch:
        dirty = ", dirty" if snapshot.git_dirty else ", clean"
        worktree = ", worktree" if snapshot.git_worktree else ""
        commit = f" ({snapshot.git_commit}{dirty}{worktree})" if snapshot.git_commit else ""
        lines.append(f"Git: {snapshot.git_branch}{commit}")

    # Runtimes
    if snapshot.available_runtimes:
        rt_parts = [f"{name} {ver}" for name, ver in sorted(snapshot.available_runtimes.items())]
        lines.append(f"Runtimes: {', '.join(rt_parts)}")

    # Filesystem summary
    if snapshot.notable_files:
        top_files = snapshot.notable_files[:8]
        extras = ""
        if snapshot.file_count or snapshot.directory_count:
            extras = f" ({snapshot.file_count} files, {snapshot.directory_count} dirs)"
        lines.append(f"Notable: {', '.join(top_files)}{extras}")

    # Packages summary
    if snapshot.installed_packages:
        pkg_count = len(snapshot.installed_packages)
        lockfile_note = f" ({', '.join(snapshot.lockfiles_found)})" if snapshot.lockfiles_found else ""
        lines.append(f"Packages: {pkg_count} top-level{lockfile_note}")

    return "\n".join(lines)


def render_full_json(snapshot: EnvironmentSnapshot) -> str:
    """Full JSON serialization for artifact persistence."""
    return json.dumps(snapshot.to_dict(), indent=2, sort_keys=True)
